### PR TITLE
NNC depthwise conv2d implementation

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55213 Integrate NNC conv2d with fuser
* #55025 [nnc] Move device type checks from isSupported to typesAreSupported
* **#54920 NNC depthwise conv2d implementation**

Add a depthwise convolution implementation and reasonably good
schedules for 3x3 stride=1,2.

Differential Revision: [D27413745](https://our.internmc.facebook.com/intern/diff/D27413745/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27413745/)!

### CODE REVIEW

File: test/cpp/tensorexpr/test_conv.cpp [hunk 1]
```diff
@@ -10,6 +12,44 @@ namespace jit {
...
+  te::Placeholder input("input", te::kFloat, {N, C, H, W});
+  te::Placeholder weight("weight", te::kFloat, {K, CperG, R, S});
+  te::Placeholder bias("bias", te::kFloat, {K});
+  te::Tensor* output = te::conv2d_depthwise(
+      input.handle(), weight.handle(), bias.handle(), kStride, kPad, kGroups);
+
+  te::LoopNest loop({output});
+  loop.simplify();
+  loop.prepareForCodegen();
+  te::LLVMCodeGen cg(loop.root_stmt(), {input, weight, bias, output});
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_conv.cpp:37]: This might need `ifdef TORCH_ENABLE_LLVM`.

File: torch/csrc/jit/tensorexpr/operators/conv2d.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <torch/csrc/jit/tensorexpr/loopnest.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/operators/conv2d.h:3]: Nit: this include could be moved to the cpp file.

File: torch/csrc/jit/tensorexpr/operators/conv2d.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#include <torch/csrc/jit/tensorexpr/tensor.h>
+
+namespace torch::jit::tensorexpr {
+
+Tensor* conv2d_depthwise(
+    BufHandle input,
+    BufHandle weight,
+    BufHandle bias,
+    int stride,
+    int pad,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/operators/conv2d.cpp:11]: Do we plan to support different W and H strides and pads?

File: torch/csrc/jit/tensorexpr/operators/conv2d.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    int groups) {
+  TORCH_INTERNAL_ASSERT(input.ndim() == 4);
+  TORCH_INTERNAL_ASSERT(weight.ndim() == 4);
+
+  auto const& N = immediateAs<int>(input.dim(0));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/operators/conv2d.cpp:16]: Should we check if dims are actually immediates or are we ok with an assertion in `immediateAs<int>`?

File: torch/csrc/jit/tensorexpr/operators/conv2d.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      },
+      {{C / groups, "c"}, {R, "r"}, {S, "s"}});
+
+  LoopNest nest({conv});
+
+  constexpr int kLoopH = 2, kLoopW = 3;
+  if (R == 3 && stride == 2 && pad == 1) {
+    For *head, *tail;
+    auto loops = nest.getLoopStmtsFor(conv);
+    nest.sliceHead(loops[kLoopW], 2, &head, &tail);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/operators/conv2d.cpp:66]: Note for myself: we probably should add a version of `sliceHead` that doesn't require the last two arguments.

File: torch/csrc/jit/tensorexpr/expr.h [hunk 6]
```diff
@@ -10,6 +10,8 @@
 #include <torch/csrc/jit/tensorexpr/mem_arena.h>
 #include <torch/csrc/jit/tensorexpr/types.h>
 
+#include <ATen/core/functional.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:13]: Is it only for `fmap`? You could use `ExprVectorToExprHandleVector` if that doesn't look too ugly :)

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me! I have only minor comments.

File: torch/csrc/jit/tensorexpr/expr.h (hunk 6 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/expr.h:13]: Yeah just for fmap.  I have this allergy to super long function names :-p

File: torch/csrc/jit/tensorexpr/operators/conv2d.cpp (hunk 3 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/operators/conv2d.cpp:11]: I don't have a strong opinion on this, was just curious :)

File: torch/csrc/jit/tensorexpr/operators/conv2d.cpp (hunk 4 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/operators/conv2d.cpp:16]: yeah, good point, `immediateAs` will give a really opaque error message like "UNSUPPORTED DTYPE" (which is maybe a separate thing to fix but not quite the fault of this diff :-p ).  I'll add an explicit check.

File: test/cpp/tensorexpr/test_conv.cpp (hunk 1 above)
Comment (asuhan) [reply test/cpp/tensorexpr/test_conv.cpp:38]: Is there any other reason for it except the slowness of the IR evaluator?

**Review by asuhan — APPROVED** [review]

File: test/cpp/tensorexpr/test_conv.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply test/cpp/tensorexpr/test_conv.cpp:38]: It's probably the reason of the linker errors on this PR: if pytorch is built without LLVM, `te::LLVMCodeGen` will be an undefined symbol.

**Pull request closed without merging**